internal/tiger/mcp/db_tools.go

```diff
@@ -157,59 +159,149 @@ func (s *Server) handleDBExecuteQuery(ctx context.Context, req *mcp.CallToolRequ
+	// Extended protocol sends parameters separately but doesn't support
+	// multi-statement queries. This means we don't support multi-statement
+	// queries with parameters (pgx will return an error for them when using
+	// QueryExecModeDescribeExec). See [pgx.QueryExecMode] for details.
+	if len(input.Parameters) > 0 {
+		// Use extended protocol to send parameters separately (more secure,
+		// but doesn't support multi-statement queries).
+		connConfig.DefaultQueryExecMode = pgx.QueryExecModeDescribeExec
+	} else {
+		// Use simple protocol to support multi-statement queries when no
+		// parameters are given.
+		connConfig.DefaultQueryExecMode = pgx.QueryExecModeSimpleProtocol
+	}
+
 	// Connect to database
-	conn, err := pgx.Connect(queryCtx, details.String())
+	conn, err := pgx.ConnectConfig(queryCtx, connConfig)
 	if err != nil {
-		return nil, DBExecuteQueryOutput{}, fmt.Errorf("failed to connect to database: %w", err)
+		return nil, DBExecuteQueryOutput{}, err
 	}
 	defer conn.Close(context.Background())
 
 	// Execute query and measure time
 	startTime := time.Now()
-	rows, err := conn.Query(queryCtx, input.Query, util.ConvertSliceToAny(input.Parameters)...)
-	if err != nil {
-		return nil, DBExecuteQueryOutput{}, fmt.Errorf("query execution failed: %w", err)
+
+	// Queue the query. When using QueryExecModeSimpleProtocol (no parameters),
+	// it's valid to queue a single multi-statement SQL query as the batch.
+	// See the [pgx.Batch.Queue] documentation for details. When using
+	// QueryExecModeDescribeExec (with parameters), queueing a multi-statement
+	// query here will result in an error when executing it below.
+	batch := &pgx.Batch{}
+	batch.Queue(input.Query, util.ConvertSliceToAny(input.Parameters)...)
+
+	br := conn.SendBatch(queryCtx, batch)
+	defer br.Close()
+
+	// Process all result sets, keeping only the final one
+	var finalResult resultSet
+	for {
+		rows, err := br.Query()
+		if err != nil {
+			// Check if we've reached the final result set and stop iteration.
+			// NOTE: It would be nice if there was a real sentinel error type
+			// we could check here instead of comparing error strings, but pgx
+			// doesn't expose one. We will just need to verify that the error
+			// message doesn't change when we update the pgx dependency.
+			if err.Error() == "no more results in batch" {
+				break
+			}
+			return nil, DBExecuteQueryOutput{}, err
+		}
+
+		// Process this result set
+		result, err := processResultSet(conn, rows)
+		if err != nil {
+			return nil, DBExecuteQueryOutput{}, err
+		}
+
+		// Save this result set as the current "final" one
+		finalResult = result
+	}
+
+	if err := br.Close(); err != nil {
+		return nil, DBExecuteQueryOutput{}, err
 	}
+
+	// Build output from the final result set
+	output := DBExecuteQueryOutput{
+		Columns:       finalResult.columns,
+		Rows:          finalResult.rows,
+		RowsAffected:  finalResult.rowsAffected,
+		ExecutionTime: time.Since(startTime).String(),
```

Comment:
This includes the time to fetch/process the results. Not sure if there is a way to measure the actual execution time (in the database), but perhaps this should have a different name?

Reply:
I don't think there is a way to measure the query time took from the perspective of the database, so I'm not really sure it matters? There also isn't _really_ a clear distinction between "time to fetch/process the results" and "time (in the database)", since a client that's slow to fetch/process results will generate backpressure on the TCP connection, which goes all the way back to the database and can prevent it from continuing to scan results from disk pages and send them. In other words, you can make a query hang "in the database" by just not fetching results from the network buffer as fast as the database is trying to send them.